kernel/src/engine/arrow_expression.rs

```diff
@@ -526,6 +526,87 @@ impl ExpressionHandler for ArrowExpressionHandler {
             output_type,
         })
     }
+
+    fn create_one(&self, schema: SchemaRef, expr: &Expression) -> DeltaResult<Box<dyn EngineData>> {
+        if let Expression::Struct(child_exprs) = expr {
+            if schema.len() != child_exprs.len() {
+                return Err(Error::Generic(format!(
+                    "Schema has {} top-level fields, but struct expr has {} children",
+                    schema.len(),
+                    child_exprs.len()
+                )));
+            }
+
+            let arrays: Vec<ArrayRef> = schema
+                .fields()
+                .zip(child_exprs.iter())
+                .map(|(field, child_expr)| create_single_row_array(child_expr, field))
+                .collect::<Result<Vec<_>, Error>>()?;
+
+            let record_batch = RecordBatch::try_new(Arc::new(schema.as_ref().try_into()?), arrays)?;
+            Ok(Box::new(ArrowEngineData::new(record_batch)))
+        } else {
+            Err(Error::generic(
+                "ArrowExpressionHandler::create_one() requires a top-level struct expression",
+            ))
+        }
+    }
+}
+
+fn create_single_row_array(expr: &Expression, field: &StructField) -> DeltaResult<ArrayRef> {
+    match expr {
+        // simple case: for literals, just create a single-row array and ensure the data types match
+        Expression::Literal(scalar) => {
+            let array = scalar.to_array(1)?;
+            // TODO(zach): we could do better and cast here instead of always failing
+            ensure_data_types(field.data_type(), array.data_type(), true)?;
+            Ok(array)
+        }
+        // recursive case: for struct expressions, build a struct array by recursing into each child
+        Expression::Struct(child_exprs) => {
+            // co-traverse the expression and schema: we expect the data type to be struct, error
+            // otherwise.
+            match field.data_type() {
+                DataType::Struct(struct_type) => {
+                    if struct_type.len() != child_exprs.len() {
+                        return Err(Error::Generic(format!(
+                            "Schema struct field has {} children, but expression has {} children",
+                            struct_type.len(),
+                            child_exprs.len()
+                        )));
+                    }
+
+                    let child_arrays: Vec<ArrayRef> = struct_type
+                        .fields()
+                        .zip(child_exprs.iter())
+                        .map(|(subfield, subexpr)| create_single_row_array(subexpr, subfield))
+                        .collect::<Result<Vec<ArrayRef>, Error>>()?;
+
+                    let arrow_fields = struct_type
+                        .fields()
+                        .map(|f| ArrowField::try_from(f))
+                        .collect::<Result<Vec<ArrowField>, ArrowError>>()?;
```

Comment:
some simplifications :)
```suggestion
                    let child_arrays = struct_type
                        .fields()
                        .zip(child_exprs)
                        .map(|(subfield, subexpr)| create_single_row_array(subexpr, subfield))
                        .try_collect()?;

                    let arrow_fields = struct_type
                        .fields()
                        .map(ArrowField::try_from)
                        .try_collect()?;
```